Epic fail

alikins opened the pull request:
##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->

 - New Module Pull Request


##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
lib/ansible/modules/utility/logic/epic_fail.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.3.0 (epic_fail 93e5f21829) last updated 2017/02/08 17:08:03 (GMT -400)
  config file = /home/adrian/.ansible.cfg
  configured module search path = [u'/home/adrian/ansible/my-modules', u'/home/adrian/ansible/lib/modules/']


```

##### SUMMARY
For testing purposes, to let a playbook trigger an assortment of normal and abnormal module failure modes.

alikins committed:
Add a epic_fail module

This module fails. It fails even more than the 'fail'
module. It even fails at failing completely because it
is a failure.

This module will never amount to anything.

alikins committed:
update, gpl, docs

mattclay commented on lib/ansible/modules/utilities/logic/epic_fail.py:
```diff
@@ -0,0 +1,161 @@
+
+    # exit sans fail_json or exit_json
+    if sys_exit:
+        sys.exit(return_code or 14)
```
CI failure due to:

```
2017-02-08 22:16:29 ============================================================================
2017-02-08 22:16:29 lib/ansible/modules/utilities/logic/epic_fail.py
2017-02-08 22:16:29 ============================================================================
2017-02-08 22:16:29 ERROR: sys.exit() call found. Should be exit_json/fail_json
```

Since this is intentional, we'll need to figure out a way to permit this without disabling the module validator entirely.

sivel replied:
A few ideas:

1. Assign sys.exit to a var and call that
1. Add some form of noqa to validate-modules
1. Implement pep like codes in validate-modules that can allow ignores per file

alikins committed:
change name of the passed in return data output->return

alikins committed:
updating

alikins committed:
add a wip test role

alikins committed:
simplify a bit, make exit_json/fail_json failure_modes

need to update docs tests to indicate return code should
be specified in return_code or return_code_updates

alikins committed:
moving intg test to roles for now

mattclay commented:
@alikins If you include `EXPECTED FAILURE` in the name of a task which is expected to fail, then the integration tests will pass when it fails (assuming you handle the failure in the playbook).

Example: https://github.com/ansible/ansible/blob/devel/test/integration/targets/blocks/main.yml#L18

alikins committed:
document output_list, add exception_type, exception_args

alikins committed:
add epic_fail to validate-modules skip list

epic_fail uses sys.exit() directly to simulate a bad
module, so validate-modules will always fail.

alikins committed:
more tests, mv tests to targets/

alikins committed:
add aliases

mattclay commented:
@alikins Is this PR something you'd still like to work on? If so, the new consolidated ignore system for sanity tests may make it easier to work with, since nearly all tests support skip and ignore entries now.

mattclay commented:
Closed, since new modules should go into a collection instead of ansible-core.
